<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [13]:
import pandas as pd
import psycopg2

In [15]:
connection = psycopg2.connect(
   dbname=DBNAME,
   user=USER,
   host=HOST,
   password=PASSWORD,
   port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [16]:
# текст запроса
query = f'''select count(id) количество_вакансий_в_базе 
           from VACANCIES
                '''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\1211064281.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


количество_вакансий_в_базе
0                       49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [66]:
query = f'''select count(id) количество_работодателей_в_базе
           from EMPLOYERS'''
df = pd.read_sql_query(query, connection)
df     # текст запроса

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\2783329365.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


количество_работодателей_в_базе
0                            23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [68]:
query = f'''select count(id) количество_регионов
           from AREAS'''
df = pd.read_sql_query(query, connection)
df  # текст запроса

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\1335649776.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


количество_регионов
0                 1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [70]:
query = f'''select count(id) количество_сфер_деятельности
           from INDUSTRIES'''
df = pd.read_sql_query(query, connection)
df  # результат запроса# текст запроса

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\2203954399.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


количество_сфер_деятельности
0                           294

***

In [24]:
Вывод: В базе представлено 49197 вакансий от 23501 работодателей с 294 сферами деятельности из 1362 регионов.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [71]:
query = f''' select count(v.id) cnt, area_id area_id, a.name name_area
           from VACANCIES v
          full join areas a  on v.area_id=a.id
           group by a.name, area_id
           order by 1 desc 
          limit 5'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\1024572447.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


cnt  area_id        name_area
0  5333        1           Москва
1  2851        2  Санкт-Петербург
2  2112     1002            Минск
3  2006        4      Новосибирск
4  1892      160           Алматы

In [26]:
# результат запроса

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [72]:
query = f''' select count(distinct(v.id)) количество_вакансий
           from VACANCIES v
          where v.salary_from is not null or v.salary_to is not null'''
df = pd.read_sql_query(query, connection)
df  # текст запроса

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\1699127295.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


количество_вакансий
0                24073

Дополнительные исследования:  с каким  графиком работы больше указано вакансий с заработной платой (хотябы в одном поле) и не указано совсем 

In [73]:
query = f'''
          with  x as (
          select count(v.id) count_v_есть_ЗП, v.schedule график
           from VACANCIES v
          where v.salary_from is not null or v.salary_to is not  null
          group by v.schedule)
          select count(v.id) count_v_нет_ЗП, v.schedule график, x. count_v_есть_ЗП
          from VACANCIES v
           join x on x.график= v.schedule
          where v.salary_from is  null and v.salary_to is   null
          group by v.schedule,x. count_v_есть_ЗП
          '''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\745892999.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count_v_нет_ЗП            график  count_v_есть_ЗП
0           19047       Полный день            17377
1            1186     Гибкий график              853
2             233    Сменный график              821
3            4559  Удаленная работа             4752
4              99    Вахтовый метод              270

Выведем вакансии со сменным графиком и посомтрим какой  максимальный уровень заработной платы предлагают соискателям 

In [74]:
query = f'''
          select v.schedule, v.name, v.salary_to
         from VACANCIES v
         where v.schedule ilike '%сменн%' and v.salary_to is not null
         group by v.schedule, v.name,v.salary_to
         order by 3 desc
         limit 20
          '''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\175088280.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


schedule                                               name   
0   Сменный график                                 Разработчик Python  \
1   Сменный график       Консультант по сопровождению конфигураций 1С   
2   Сменный график            Продавец-консультант мебели Panda Terra   
3   Сменный график          Инженер технической поддержки (2-я линия)   
4   Сменный график                                   Руководитель СТО   
5   Сменный график                     Ночной системный администратор   
6   Сменный график            Middle Unity Developer в игровой проект   
7   Сменный график             Middle Серверный разработчик на mirror   
8   Сменный график       Контролер технического состояния автомобилей   
9   Сменный график  Ночной сотрудник в отдел технической поддержки...   
10  Сменный график  Сотрудник отдела технической поддержки со знан...   
11  Сменный график  Специалист по обслуживанию майнинг ферм / Сист...   
12  Сменный график  Специалист технической поддержки / Support Eng...   
13  Сменный график                     Ночной системный администратор   
14  Сменный график  Старший специалист по мониторингу кибербезопас...   
15  Сменный график  Инженер технической поддержки/системный админи...   
16  Сменный график  Ночной сотрудник отдела технической поддержки ...   
17  Сменный график                   Оператор мебельного производства   
18  Сменный график  Оператор термопластавтоматов и экструзионных л...   
19  Сменный график          Специалист/оператор технической поддержки   

    salary_to  
0      200000  
1      150000  
2      150000  
3      140000  
4      120000  
5      110000  
6      100000  
7      100000  
8      100000  
9      100000  
10     100000  
11     100000  
12     100000  
13      99362  
14      92000  
15      90000  
16      90000  
17      90000  
18      90000  
19      90000

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [75]:
query = f''' select round(avg(v.salary_from),0) avg_s_from,round(avg(v.salary_to),0) avg_s_to
           from VACANCIES v'''
df = pd.read_sql_query(query, connection)
df # текст запроса

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\2540347097.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


avg_s_from  avg_s_to
0     71065.0  110537.0

Дополнительные исследования: найдем регионы предлагающие максимальную и минимальную забаботную плату, исключая из исследования вакансии, где заработная плата не указана совсем.

Максимальные заработные платы:

In [76]:
query = f''' select max(v.salary_to) max_s_to,   a.name
  from VACANCIES v
  left join areas a on a.id=v.area_id
  group by  a.id,v.salary_to, v.salary_from
  having v.salary_to is not null and v.salary_from is not null
  order by 1 desc,2
  limit 5'''
df = pd.read_sql_query(query, connection)
df 

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\3664960273.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


max_s_to             name
0   1000000           Алматы
1   1000000           Москва
2   1000000             Сочи
3    950000           Москва
4    935176  Санкт-Петербург

Минимальные заработные платы:

In [77]:
query = f''' select min(v.salary_from) min_s_from,   a.name
  from VACANCIES v
  left join areas a on a.id=v.area_id
  group by  a.name,v.salary_from, v.salary_to
  having v.salary_from is not null and v.salary_to is not null
  order by 1,2
  limit 5
  '''
df = pd.read_sql_query(query, connection)
df 

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\1207782615.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


min_s_from       name
0          30  Челябинск
1          42    Воронеж
2          60   Смоленск
3          80        Уфа
4          80      Химки

Найдем среднюю максимальную заработную плату  для Алматы.

In [78]:
query = f'''  with x as ( select avg(v.salary_to) avg_s_to, a.name
  from VACANCIES v
  left join areas a on a.id=v.area_id
  where a.name  ilike '%Алмат%'
  group by  a.name,v.salary_to,v.salary_from
  having v.salary_to is not null and v.salary_from is not null)
  select * 
  from x
  union all
  select 
  round(avg(x.avg_s_to),0),'avg_max_to'
  from x
  order by 2
  limit 1'''
df = pd.read_sql_query(query, connection)
df 

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\3458603274.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


avg_s_to        name
0  111835.0  avg_max_to

Найдем среднюю максимальную заработную плату  для Москвы.

In [79]:
query = f'''  with x as ( select avg(v.salary_to) avg_s_to, a.name
  from VACANCIES v
  left join areas a on a.id=v.area_id
  where a.name  ilike '%москв%'
  group by  a.name,v.salary_to,v.salary_from
  having v.salary_to is not null and v.salary_from is not null)
  select * 
  from x
  union all
  select 
  round(avg(x.avg_s_to),0),'avg_max_to'
  from x
  order by 2
  limit 1'''
df = pd.read_sql_query(query, connection)
df 

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\173873307.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


avg_s_to        name
0  169929.0  avg_max_to

Найдем среднюю максимальную заработную плату  для Сочи.

In [80]:
query = f'''  with x as ( select avg(v.salary_to) avg_s_to, a.name
  from VACANCIES v
  left join areas a on a.id=v.area_id
  where a.name  ilike '%соч%'
  group by  a.name,v.salary_to,v.salary_from
  having v.salary_to is not null and v.salary_from is not null)
  select * 
  from x
  union all
  select 
  round(avg(x.avg_s_to),0),'avg_max_to'
  from x
  order by 2
  limit 1'''
df = pd.read_sql_query(query, connection)
df 

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\519548020.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


avg_s_to        name
0  152180.0  avg_max_to

Найдем среднюю максимальную заработную плату  для Санкт-Петербурга.

In [81]:
query = f'''  with x as ( select avg(v.salary_to) avg_s_to, a.name
  from VACANCIES v
  left join areas a on a.id=v.area_id
  where a.name  ilike '%санкт%'
  group by  a.name,v.salary_to,v.salary_from 
  having v.salary_to is not null and v.salary_from is not null)
  select * 
  from x
  union all
  select 
  round(avg(x.avg_s_to),0),'avg_max_to'
  from x
  order by 2
  limit 1'''
df = pd.read_sql_query(query, connection)
df 

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\3231234028.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


avg_s_to        name
0  141734.0  avg_max_to

Найдем среднюю заработную плату в Сочи.

In [82]:
query = f'''with x as(
  select round(avg(v.salary_from),0) avg_s_from,round(avg(v.salary_to),0) avg_s_to, ((avg(v.salary_from)+avg(v.salary_to))/2) total_avg
           from VACANCIES v
  left join areas a on a.id=v.area_id         
  where a.name  ilike '%соч%'
  group by v.salary_to,v.salary_from
  having v.salary_to is not null and v.salary_from is not null)
  select *
  from x
  union all
  select
  round(avg(x.avg_s_from)),round(avg(x.avg_s_to)),round(avg(x.total_avg))
  from x'''
df = pd.read_sql_query(query, connection)
df.tail(1) 

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\2912975469.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


avg_s_from  avg_s_to  total_avg
52     93505.0  152180.0   122843.0

Найдем среднюю заработную плату в Москве.

In [84]:
query = f'''with x as(
  select round(avg(v.salary_from),0) avg_s_from,round(avg(v.salary_to),0) avg_s_to, ((avg(v.salary_from)+avg(v.salary_to))/2) total_avg
           from VACANCIES v
  left join areas a on a.id=v.area_id         
  where a.name  ilike '%москв%'
  group by v.salary_to,v.salary_from
  having v.salary_to is not null and v.salary_from is not null)
  select *
  from x
  union all
  select
  round(avg(x.avg_s_from)),round(avg(x.avg_s_to)),round(avg(x.total_avg))
  from x'''
df = pd.read_sql_query(query, connection)
df.tail(1) 

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\3589570956.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


avg_s_from  avg_s_to  total_avg
283    112435.0  169929.0   141182.0

Найдем среднюю заработную плату в Алматы.

In [88]:
query = f'''with x as(
  select round(avg(v.salary_from),0) avg_s_from,round(avg(v.salary_to),0) avg_s_to, ((avg(v.salary_from)+avg(v.salary_to))/2) total_avg
           from VACANCIES v
  left join areas a on a.id=v.area_id         
  where a.name  ilike '%алмат%'
  group by v.salary_to,v.salary_from
  having v.salary_to is not null and v.salary_from is not null)
  select *
  from x
  union all
  select
  round(avg(x.avg_s_from)),round(avg(x.avg_s_to)),round(avg(x.total_avg))
  from x'''
df = pd.read_sql_query(query, connection)
df.tail(1) 

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\1550669857.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


avg_s_from  avg_s_to  total_avg
181     65641.0  111835.0    88738.0

Найдем среднюю заработную плату в Санкт-Петербурге.

In [89]:
query = f'''with x as(
  select round(avg(v.salary_from),0) avg_s_from,round(avg(v.salary_to),0) avg_s_to, ((avg(v.salary_from)+avg(v.salary_to))/2) total_avg
           from VACANCIES v
  left join areas a on a.id=v.area_id         
  where a.name  ilike '%санк%'
  group by v.salary_to,v.salary_from
  having v.salary_to is not null and v.salary_from is not null)
  select *
  from x
  union all
  select
  round(avg(x.avg_s_from)),round(avg(x.avg_s_to)),round(avg(x.total_avg))
  from x'''
df = pd.read_sql_query(query, connection)
df.tail(1) 

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\2459101508.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


avg_s_from  avg_s_to  total_avg
266     88751.0  141734.0   115242.0

Найдем среднюю заработную плату в Перми.

In [90]:
query = f'''with x as(
  select round(avg(v.salary_from),0) avg_s_from,round(avg(v.salary_to),0) avg_s_to, ((avg(v.salary_from)+avg(v.salary_to))/2) total_avg
           from VACANCIES v
  left join areas a on a.id=v.area_id         
  where a.name  ilike '%перм%'
  group by v.salary_to,v.salary_from
  having v.salary_to is not null and v.salary_from is not null)
  select *
  from x
  union all
  select
  round(avg(x.avg_s_from)),round(avg(x.avg_s_to)),round(avg(x.total_avg))
  from x'''
df = pd.read_sql_query(query, connection)
df.tail(1) 

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\631773786.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


avg_s_from  avg_s_to  total_avg
107     83554.0  130128.0   106841.0

Найдем самые  высокооплачиваемые вакансии в Перми.

In [95]:
query = f'''with x as(
  select round(avg(v.salary_from),0) avg_s_from,round(avg(v.salary_to),0) avg_s_to, ((avg(v.salary_from)+avg(v.salary_to))/2) total_avg, v.name name_v
           from VACANCIES v
  left join areas a on a.id=v.area_id         
  where a.name  ilike '%перм%'
  group by v.salary_to,v.salary_from, v.name
  having v.salary_to is not null and v.salary_from is not null)
  select *
  from x
  union all
  select
  avg(x.avg_s_from),avg(x.avg_s_to), avg(x.total_avg), x.name_v
  from x
  group by x.name_v
  order by 2 desc
  limit 5'''
df = pd.read_sql_query(query, connection)
df

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\3831720716.py:17: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


avg_s_from  avg_s_to  total_avg                             name_v
0    400000.0  800000.0   600000.0            Product manager LATOKEN
1    400000.0  800000.0   600000.0            Product manager LATOKEN
2    150000.0  450000.0   300000.0                        Программист
3    150000.0  450000.0   300000.0                        Программист
4    220000.0  350000.0   285000.0  Android-разработчик (Java/Kotlin)

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [97]:
query = f''' select distinct
           v.employment, v.schedule, count(*)
           from VACANCIES v
         -- where v.salary_from is not null or v.salary_to is not null
        group by  v.employment, v.schedule
        order by 3 desc
        '''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\2629008037.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


employment          schedule  count
0      Полная занятость       Полный день  35367
1      Полная занятость  Удаленная работа   7802
2      Полная занятость     Гибкий график   1593
3   Частичная занятость  Удаленная работа   1312
4      Полная занятость    Сменный график    940
5            Стажировка       Полный день    569
6      Полная занятость    Вахтовый метод    367
7   Частичная занятость       Полный день    347
8   Частичная занятость     Гибкий график    312
9      Проектная работа       Полный день    141
10     Проектная работа  Удаленная работа    133
11           Стажировка     Гибкий график    116
12  Частичная занятость    Сменный график    101
13           Стажировка  Удаленная работа     64
14     Проектная работа     Гибкий график     18
15           Стажировка    Сменный график     12
16     Проектная работа    Вахтовый метод      2
17     Проектная работа    Сменный график      1

Дополнительно исследуем пары с удаленной работой.

In [98]:
query = f''' select distinct
           v.employment, v.schedule, count(*)
           from VACANCIES v
         where v.schedule='Удаленная работа'
        group by  v.employment, v.schedule
        order by 3 desc
        limit 10'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\1218266699.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


employment          schedule  count
0     Полная занятость  Удаленная работа   7802
1  Частичная занятость  Удаленная работа   1312
2     Проектная работа  Удаленная работа    133
3           Стажировка  Удаленная работа     64

Посмотрим количество вакансий в разрезе графиков работы, выведем в порядке убывания

In [99]:
query = f''' with x as(
        select distinct
           v.employment, v.schedule, count(*) количество
           from VACANCIES v
        group by  v.employment, v.schedule
        order by 3 desc)
        select *
        from x
    union all
    select 'всего с графиком:', ' '||x.schedule||'', sum(x.количество) 
    from x
    group by x.schedule
    order by 1,3 desc
    limit 5'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\3451956368.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


employment           schedule  количество
0  всего с графиком:        Полный день     36424.0
1  всего с графиком:   Удаленная работа      9311.0
2  всего с графиком:      Гибкий график      2039.0
3  всего с графиком:     Сменный график      1054.0
4  всего с графиком:     Вахтовый метод       369.0

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [100]:
query = f''' select  v.experience, count(*) 
           from VACANCIES v
           group by  v.experience
           order by 2'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\2051486988.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

Вывод: Пятерка лмидеров регионов  по количеству вакансий: Москва, Санкт-Петербург, Минск, Новосибирск, Алматы. Однако, следует учитывать, что в базе есть регионы, которые не разместили не одной вакансии их 593. У вакансий со сменным графиком только 22 % без указания заработной платы. Как правило сменный график бывает на призводстве. Из 49197 вакансий 24073 вакансии с заполненной заработной платой, что составляет почти 50% (48.93%). 
Для соискателей на полный день из 36424 вакансий 52% без указанной заработной платы
Для соискателей  на удаленную работу из 9311 вакансий 49% без указанной заработной платы
Для соискателей с гибким графиком из 2039 вакансий 58% без указанной заработной платы
Для соискателей на вахту из 369 вакансий 27% без указанной заработной платы
Для соискателей со сменным графиком из 1054 вакансий 22% без указанной заработной платы
Как правило работодатель не указывает уровень заработной платы, в случае открытия новой вакансии по новому направлению работы так как  еще сам не определился с суммой оплаты. 
Сотрудники со сменных графиком, чаще всего бывают, производственные рабочие, найти ожидаемую работу им проще (указан уровень ЗП работодателем). Проведя анализ вакансий со сменным графиком и с указанной ЗП выяснилось, что максимальная Зп предлагается разработчикам Питона, консультантам по сопровождению 1С и продавцу - консультанту мебели Panda Terra( на сайте  этой компании сегодня требуются производственные рабочие со средним заработком 55-60 тысяч рублей). Для более глубокого анализа в этом направлении можно посмотреть какие работодатели из каких регионов предлагают вакансии с максимальной заработной платой. И сделать выводы про регион. 
Лидеры по предлагаемой максимальной заработной плате Алматы, Москва, Сочи, Санкт-Петербург и аутсейдеры Челябинск, Воронеж, Смоленск, Уфа, но для окончательных выводов нужен более детальный анализ.
Посмотрим среднюю максимальную заработную плату в этих городах не беря в расчет вакансии где поля минимальной и максимальной заработной платы не заполнены.
средняя максимальная заработная плата:
 в Алматы 111835 рублей
 в Москве 169929 рубль
 в Сочи 152180 рублей
 в Санкт-Петербурге 141734 рублей
 В итоге средняя максимальная заработная плата в Алматы ниже Москвы, Сочи и Санкт-Петербурга, что подтверждает, что выводы по макс и минимуму делать не корректно, нужно более детально исследовать, смотреть среднюю, с учетом нулевых значений.

 Найдем среднюю заработную плату в вышеназванных городах:
в Алматы 88738.0 рублей
 в Москве 138728.0 рубль
 в Сочи 122843 рублей
 в Санкт-Петербурге 115242.0 рублей 
 Рейтинг средних заработных плат совпал со средними максимальными.
 Дополнительно исследовала среднюю заработную плату в своем городе Перми, она составила 106841.0, могу сказать по своему предприятию - у нас ниже даже нижней границы. Самая высокооплачиваемая в нашем городе вакансия Product manager LATOKEN
 При рассмотении пар тип графика и тип трудоустройсва на втором месте в списке  пара "полная занятость-удаленная работа", думаю до пандемии данное сочетание было значительно ниже в рейтинге.В существующей базе удаленая работа стоит в паре с полной занятостью, частичной занятостью, проектной работой и стажировкой. График с удаленной работой стоит на втором месте по количеству вакансий.
 Наиболее востребованным опытом работы является "от года до трех лет", уровень заработной платы у такой категории вакансий ниже, чем у вакансий  "от 3 до 6 лет" и "более 6 лет". Работодатели предпочитают взять на работу сотрудника с минимальным опытом, но в то же время имеющим навыки работы, чем обучать не опытного сотрудника или платить боле высокую заработную плату оытному.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [101]:
query = f''' select  v.employer_id, count(*) количество_вакансий, e.name
           from VACANCIES v
           join employers e on v.employer_id=e.id
           group by v.employer_id, e.name
           order by 2 desc
           limit 5'''
df = pd.read_sql_query(query, connection)
df  # текст запроса

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\2408315230.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


employer_id  количество_вакансий           name
0         1740                 1933         Яндекс
1         2748                  491     Ростелеком
2        78638                  444       Тинькофф
3         3529                  428           СБЕР
4        39305                  331  Газпром нефть

Дополнительно запросим  сферы деятельности этих компаний.

In [102]:
query = f''' select 
v.employer_id, count(*) количество_вакансий, e.name, i.name сфера_деятельности
from VACANCIES v
join employers e on v.employer_id=e.id
join employers_industries e_i on e.id=e_i.employer_id
join industries i on i.id=e_i.industry_id
group by v.employer_id, e.name, i.name
order by 2 desc
limit 14'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\1725323962.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


employer_id  количество_вакансий           name   
0          1740                 1933         Яндекс  \
1          2748                  491     Ростелеком   
2          2748                  491     Ростелеком   
3          2748                  491     Ростелеком   
4          2748                  491     Ростелеком   
5         78638                  444       Тинькофф   
6         78638                  444       Тинькофф   
7         78638                  444       Тинькофф   
8         78638                  444       Тинькофф   
9         78638                  444       Тинькофф   
10         3529                  428           СБЕР   
11        39305                  331  Газпром нефть   
12        39305                  331  Газпром нефть   
13        39305                  331  Газпром нефть   

                                   сфера_деятельности  
0   Интернет-компания (поисковики, платежные систе...  
1                                  Интернет-провайдер  
2                                 Фиксированная связь  
3                                     Мобильная связь  
4                                Оптоволоконная связь  
5                          Коллекторская деятельность  
6   Системная интеграция,  автоматизации технологи...  
7                                                Банк  
8                 Разработка программного обеспечения  
9   Интернет-компания (поисковики, платежные систе...  
10                                               Банк  
11        Нефтепереработка, нефтехимия (производство)  
12       ГСМ, топливо (продвижение, оптовая торговля)  
13                  ГСМ, топливо (розничная торговля)

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [103]:
query = f''' select
count(v.id) cnt_вакансии, count(e.id) cnt_работодателя, a.name регион
from areas a
left  join vacancies v on a.id=v.area_id
left join employers e on a.id=e.area
where v.id is null 
group by a.name
order by 1,2 desc
limit 20'''
df = pd.read_sql_query(query, connection)
df  # текст запроса

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\938725390.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


cnt_вакансии  cnt_работодателя                 регион
0              0               410                 Россия
1              0               207              Казахстан
2              0                75     Московская область
3              0                19     Краснодарский край
4              0                18     Ростовская область
5              0                18               Беларусь
6              0                17            Азербайджан
7              0                16  Нижегородская область
8              0                16   Республика Татарстан
9              0                15             Узбекистан
10             0                10   Белгородская область
11             0                10               Сингапур
12             0                 9      Калужская область
13             0                 9  Ленинградская область
14             0                 7    Воронежская область
15             0                 7      Самарская область
16             0                 6      Кировская область
17             0                 6             Кыргызстан
18             0                 6              Австралия
19             0                 6    Ставропольский край

Найдем сколько всего работодателей не указали вакансии.


In [104]:
query = f''' select
count(e.id) 
from areas a
left  join vacancies v on a.id=v.area_id
left join employers e on a.id=e.area
where v.id is null '''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\248209473.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0   1148

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [105]:
query = f''' select
count(distinct v.area_id) count_id_региона_вакансии,  e.name работодатель
from vacancies v
left join employers e on v.employer_id=e.id
group by e.id, e.name
order by 1 desc
limit 20
'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\2141501062.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count_id_региона_вакансии                работодатель
0                         181                      Яндекс
1                         152                  Ростелеком
2                         116                  Спецремонт
3                          88      Поляков Денис Иванович
4                          71                   ООО ЕФИН 
5                          63                  Совкомбанк
6                          55                         МТС
7                          49  ЭФКО, Управляющая компания
8                          48                Почта России
9                          48                        КРОН
10                         46                       MCORE
11                         46                 ИК СИБИНТЕК
12                         45             Сбербанк-Сервис
13                         44                       ANCOR
14                         43                    Тинькофф
15                         43                  Первый Бит
16                         41                       АЛНАС
17                         36                         IBS
18                         36                        Ozon
19                         35                       Авито

Запросим сферы деятельности этих компаний.

In [106]:
query = f''' select
count(distinct v.area_id) count_id_региона_вакансии, e.id id_работодателя, e.name работодатель, i.name 
from vacancies v
left join employers e on v.employer_id=e.id
left join employers_industries ei on e.id=ei.employer_id
left join industries i on ei.industry_id=i.id
group by e.id, e.name, i.name
order by 1 desc
limit 20 
'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\4076114513.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count_id_региона_вакансии  id_работодателя                работодатель   
0                         181             1740                      Яндекс  \
1                         152             2748                  Ростелеком   
2                         152             2748                  Ростелеком   
3                         152             2748                  Ростелеком   
4                         152             2748                  Ростелеком   
5                         116          5724811                  Спецремонт   
6                          88          5130287      Поляков Денис Иванович   
7                          71          3682876                   ООО ЕФИН    
8                          63             7944                  Совкомбанк   
9                          55             3776                         МТС   
10                         49            53797  ЭФКО, Управляющая компания   
11                         49            53797  ЭФКО, Управляющая компания   
12                         49            53797  ЭФКО, Управляющая компания   
13                         48          3776815                        КРОН   
14                         48             4352                Почта России   
15                         48          3776815                        КРОН   
16                         46           197135                 ИК СИБИНТЕК   
17                         46           622121                       MCORE   
18                         45          1473866             Сбербанк-Сервис   
19                         45          1473866             Сбербанк-Сервис   

                                                 name  
0   Интернет-компания (поисковики, платежные систе...  
1                                     Мобильная связь  
2                                Оптоволоконная связь  
3                                 Фиксированная связь  
4                                  Интернет-провайдер  
5                                                None  
6                                                None  
7                                                None  
8                                                Банк  
9                                     Мобильная связь  
10              Масложировая продукция (производство)  
11  Масложировая продукция (продвижение, оптовая т...  
12                      Несетевая розница, мелкий опт  
13                Разработка программного обеспечения  
14                      Курьерская, почтовая доставка  
15  Системная интеграция,  автоматизации технологи...  
16  Системная интеграция,  автоматизации технологи...  
17  Системная интеграция,  автоматизации технологи...  
18                                               Банк  
19  Системная интеграция,  автоматизации технологи...

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [107]:
query = f''' select
count(e.id) count_работодателей
from employers e
left join employers_industries ei on e.id=ei.employer_id 
where ei.industry_id is null

'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\1525474290.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count_работодателей
0                 8419

Выведем самых крупных работодателей по количеству вакансий , которые не указали свою сферу деятельности.

In [108]:
query =  f''' select
count(v.id) count_вакансии, e.id id_работодателя, e.name работодатель, i.name сфера_деятельности
from vacancies v
left join employers e on v.employer_id=e.id
left join employers_industries ei on e.id=ei.employer_id
left join industries i on ei.industry_id=i.id
where i.id is null
group by e.id, e.name, i.name
order by 1 desc
limit 20 
'''
df = pd.read_sql_query(query, connection)
df  


C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\2027715127.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count_вакансии  id_работодателя                          работодатель   
0              151          5724811                            Спецремонт  \
1              127          5130287                Поляков Денис Иванович   
2              122          4649269              Иннотех, Группа компаний   
3               72          5920492                     DNS Головной офис   
4               71          3682876                             ООО ЕФИН    
5               69           213349                                 АЛНАС   
6               68          4914081                                Диплей   
7               54          4029257        РУСАЛ, Центр подбора персонала   
8               47          5846248                            Employcity   
9               46          3144945  Ростелеком Информационные Технологии   
10              44          5884778                                  Эр-1   
11              42          4418001      АЛРОСА Информационные технологии   
12              39          5912977                                  Кама   
13              34          3635947                    Pride Games Studio   
14              34          1689259                               JCat.ru   
15              34          3890345                      Reelmotion games   
16              32          4446629                               4people   
17              30          6171199                            A4.finance   
18              30          3191643                                 ЦИФРА   
19              29           909194                                  Нэти   

   сфера_деятельности  
0                None  
1                None  
2                None  
3                None  
4                None  
5                None  
6                None  
7                None  
8                None  
9                None  
10               None  
11               None  
12               None  
13               None  
14               None  
15               None  
16               None  
17               None  
18               None  
19               None

Найдем ТОП 5  регионов, где работодатели не указали сферу деятельности. 

In [109]:
query = f''' select
count(e.id) count_работодателей, e.area,a.name
from employers e
left join employers_industries ei on e.id=ei.employer_id 
join areas a on e.area=a.id
where ei.industry_id is null
group by e.area, a.name
order by 1 desc
limit 5
'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\2032792932.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count_работодателей  area             name
0                 1997     1           Москва
1                  714     2  Санкт-Петербург
2                  254  1002            Минск
3                  227     3     Екатеринбург
4                  207     4      Новосибирск

In [43]:
# результат запроса

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [110]:
query = f''' select
e.name name_работодателя
from employers e
left join employers_industries ei on e.id=ei.employer_id 
group by e.id
having count(ei.industry_id)=4
order by 1
limit 1 offset 2


'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\587608072.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


name_работодателя
0              2ГИС

Исследуем какое найбольшее количество сфер деятельности указывают комапнии.

In [111]:
query = f''' select count(ei.industry_id) количество_сфер_деятельности , e.name name_работодателя
from employers e
left join employers_industries ei on e.id=ei.employer_id 
group by e.id
order by 1 desc
limit 10
'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\2903682045.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


количество_сфер_деятельности   
0                             5  \
1                             5   
2                             5   
3                             5   
4                             5   
5                             5   
6                             5   
7                             5   
8                             5   
9                             5   

                                   name_работодателя  
0                                         Darkstudio  
1                                            Флагман  
2                                           АЗИЯТЕКС  
3                                        Альфа Строй  
4                                             ГК АСБ  
5                                               ИСКТ  
6                                       АМС-ЛОГИСТИК  
7                                        Биоэкология  
8                                Сернурский сырзавод  
9  ООО «Мицубиси Электрик (РУС)» / Mitsubishi Ele...

Посмотрим сколько сфер деятельности указывают компании с самым большим количеством вакансий .

In [112]:
query = f''' with  x as (
select count(ei.industry_id) количество_сфер, e.name name_работодателя, e.id id_работодателя
from employers e
left join employers_industries ei on e.id=ei.employer_id 
group by e.name, e.id
order by 1 desc)
select
x.количество_сфер,x.name_работодателя,count(v.id) количество_вакансий
from x
left join VACANCIES v on v.employer_id=x.id_работодателя
group by x.id_работодателя, x.количество_сфер,x.name_работодателя
order by 3 desc
limit 10

'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\4053356870.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


количество_сфер      name_работодателя  количество_вакансий
0                1                 Яндекс                 1933
1                4             Ростелеком                  491
2                5               Тинькофф                  444
3                1                   СБЕР                  428
4                3          Газпром нефть                  331
5                1            ИК СИБИНТЕК                  327
6                1                    МТС                  292
7                2                DataArt                  247
8                1  Совкомбанк Технологии                  204
9                1             Первый Бит                  176

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [113]:
query = f''' select
 count(ei.employer_id) 
from employers_industries ei
join industries  ind on ind.id=ei.industry_id
where ind.name like  '%Разработка программного обеспечения%'
'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\1287327641.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0   3553

Посмотрим рейтинг по убыванию сфер деятельности компании.

In [114]:
query = f''' select
 count(ei.employer_id), ind.name  сфера_деятельности
from employers_industries ei
join industries  ind on ind.id=ei.industry_id
group by ind.name
order by 1 desc
'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\664513844.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count                                 сфера_деятельности
0     3553                Разработка программного обеспечения
1     2993  Системная интеграция,  автоматизации технологи...
2     1675  Интернет-компания (поисковики, платежные систе...
3      798  Маркетинговые, рекламные, BTL, дизайнерские, E...
4      662                              Консалтинговые услуги
..     ...                                                ...
289      4        Лифтовое хозяйство (монтаж, сервис, ремонт)
290      4                                  Ритуальные услуги
291      3              Ботанический сад, зоопарк, заповедник
292      3                                          Автошкола
293      1             Ассоциация в сфере культуры, искусства

[294 rows x 2 columns]

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [116]:

import urllib.parse as req

path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров')
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)

url

'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8#%D0%A1%D0%BF%D0%B8%D1%81%D0%BE%D0%BA_%D0%B3%D0%BE%D1%80%D0%BE%D0%B4%D0%BE%D0%B2-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D0%BE%D0%B2'

In [117]:
import urllib.parse as req

path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров')
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)
cities=tuple(pd.read_html(url)[1]['Город'])
#cities = tuple(pd.read_html(url)[1]['Город'].items)
#cities=', '.join(cities)
cities

('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Красноярск',
 'Челябинск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Краснодар',
 'Омск',
 'Воронеж',
 'Пермь',
 'Волгоград')

In [118]:
query = f'''with x as( select
 a.name город, count(v.id) count_вакансий
from employers e
join vacancies v on e.id=v.employer_id
join areas a on a.id=v.area_id
where e.name ='Яндекс' and a.name in {cities}
group by a.name
 )

select *
from x
union all 
select
'total', sum(x.count_вакансий)
from x
'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\1650217648.py:17: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


город  count_вакансий
0         Волгоград            24.0
1           Воронеж            32.0
2      Екатеринбург            39.0
3            Казань            25.0
4         Краснодар            30.0
5        Красноярск            23.0
6            Москва            54.0
7   Нижний Новгород            36.0
8       Новосибирск            35.0
9              Омск            21.0
10            Пермь            25.0
11   Ростов-на-Дону            25.0
12           Самара            26.0
13  Санкт-Петербург            42.0
14              Уфа            26.0
15        Челябинск            22.0
16            total           485.0

***

Выводы: Топ-5 работодателей: Яндекс, Ростелеком, Тинькофф, СБЕР,Газпром нефть. При рассмотрении сфер деятельности этих компаний, только Газпром нефть является энергетическим предприятием. Остальные компании из списка относятся к  интернет компаниям, финансовым компаниям, компаниям из сферы IT.  Хоть и все они предстаавляют разные отрасли, но являются ключевыми игроками на рынке и оказывают влияние на экономику России.
 Всего 1148 работодателей, имеющихся в базе, не указали вакансии. Например 410 работодателей в России не указали вакансии, это может говорить и о том, что эти работодатели уже нашли сотрудников как и то, что при заполнении анкеты не верно указали данные или приостановили деятельность.
Больше всего регионов, в которых один работодатель ищет сотрудников, у компани Яндекс.  Они разместили вакансии в 181 городе. Дополнительно для анализа написала запрос выгружающий  сферу деятельности этих компаний. В основном это компании связи, но есть и те, что не указали сферу деятельности. 
Некоторые компании не указывают сферу деятельности из соображений конфиденциальности, по причине разработки нового продукта , который преждевременно афишировать.По причине  разнообразных сфер деятельности, по причине привлечения сотрудников из разных сфер, чтобы не ограничивать соискателей указанными сферами. Причины могут быть разные, вплоть до не желания заполнить это графу кадровым работником, подающим объявление.  
8419 работодателей из базы   не указали сферу деятельности, топ 5  орегионов без сферы деятельности это : Москва, Санкт-Петербург, Минск, Екатеринбург и Новосибирск.
Самое большое количество сфер деятельности указанное работодателями пять. Компания с самым большим количеством вакансий- Яндекс указала одну сферу деятельности, в то же время Ростелеком и Тинкоф указали по 4 и 5 сфер деятельности.
Из 49197 вакансий процент сферы деятельности "Разработка программного обеспечения" составляет 7%, что является самым многочисленным, далее идут системная интеграция, автоматизации технологических и бизнес-процессов предприятия, ИТ., затем итнергет-компания и так далее. Преимущественное большинство компаний публикующее вакансии  из сферы ИТ. 
Самая крупная по размещению вакансий  компания Яндекс размещает 485 вакансий в городах миллионниках, всего компания размещает 1933 вакансии,остальные приходятся на города с населением меньше миллиона жителей. 

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [119]:
query = f''' select
count(*)
from vacancies v
where v.name ilike  '%data%' or v.name ilike '%данн%'
'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\1176517903.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0   1771

Запросим количество вакансий относящиеся к продажам

In [125]:
query = f''' select
count(*)
from vacancies v
where (v.name ilike  '%прода%' or v.name ilike '%маркет%' or v.name ilike '%sale%')
'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\789302597.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0   1231

Запросим количество вакансий включающих юридическую работу.

In [126]:
query = f'''select
count(*)
from vacancies v
where v.name ilike  '%юрис%' or v.name ilike '%lawye%'
'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\3860749594.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0      9

Запросим количество вакансий включающих запрос работников бухгалтерии.

In [127]:
query = f'''select
count(*)
from vacancies v
where v.name ilike  '%бух%'
'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\2031906817.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0     97

Запросим количество вакансий включающих запрос работников экономистов.

In [128]:
query = f'''select
count(*)
from vacancies v
where v.name ilike  '%эконо%'
'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\3824508436.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0     74

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [131]:
query = f''' select
count(*)
from vacancies v
where(v.name ilike '%data scientist%' 
or v.name ilike '%data science%' 
or v.name ilike '%data science%'
 or v.name ilike '%исследователь данных%'
 or( v.name like '%ML%' and v.name not like '%HTML%')
 or v.name ilike '%machine learning%'
 or v.name ilike '%машинн%обучен%')
 and ( v.name  ilike '%junior%'  OR v.experience ILIKE '%Нет опыта%'
    OR v.employment ILIKE '%Стажировка%' )
    
'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\1771355733.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [132]:
query = f''' select
count(*)
from vacancies v
where(v.name ilike '%data scientist%' 
or v.name ilike '%data science%' 
or v.name ilike '%data science%'
 or v.name ilike '%исследователь данных%'
 or( v.name like '%ML%' and v.name not like '%HTML%')
 or v.name ilike '%machine learning%'
 or v.name ilike '%машинн%обучен%')
and (v.key_skills ilike '%SQL%' or  v.key_skills ilike '%postgres%') 
    
'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\3855629604.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0    201

Найдем сколько всего вакасий  для DS.

In [133]:
query = f''' select
count(v.id)
from vacancies v
where(v.name ilike '%data scientist%' 
or v.name ilike '%data science%' 
or v.name ilike '%data science%'
 or v.name ilike '%исследователь данных%'
 or( v.name like '%ML%' and v.name not like '%HTML%')
 or v.name ilike '%machine learning%'
 or v.name ilike '%машинн%обучен%')

    
'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\3224711420.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0    480

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [137]:
query = f''' select
count(*)
from vacancies v
where(v.name ilike '%data scientist%' 
or v.name ilike '%data science%' 
or v.name ilike '%data science%'
 or v.name ilike '%исследователь данных%'
 or( v.name like '%ML%' and v.name not like '%HTML%')
 or v.name ilike '%machine learning%'
 or v.name ilike '%машинн%обучен%')
and v.key_skills ilike '%Python%' 
    
'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\324929793.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [134]:
query = f''' with x as (select

v.key_skills, ((length(v.key_skills)-length(replace(v.key_skills,CHR(9),'')))::integer +1) sk_3
from vacancies v
where(v.name ilike '%data scientist%' 
or v.name ilike '%data science%' 
or v.name ilike '%data science%'
 or v.name ilike '%исследователь данных%'
 or( v.name like '%ML%' and v.name not like '%HTML%')
 or v.name ilike '%machine learning%'
 or v.name ilike '%машинн%обучен%'))
 select
 *
 from x
 union  all
select
 'среднее количество навыков',round(avg(x.sk_3),2)
from x
    
'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\1016521007.py:21: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


key_skills  sk_3
0            Python\tSQL\tРабота в команде\tOpenCV\tML  5.00
1    Oracle Pl/SQL\tБазы данных\tMachine Learning\t...  9.00
2    Python\tOpenCV\tTensorflow\tComputer Vision\tН...  8.00
3                       Python\tFlask\tLinux\tAWS\tGit  5.00
4     Python\tSQL\tNLP\tScikit-learn\tMachine Learning  5.00
..                                                 ...   ...
476           Python\tScikit-learn\tPandas\tGit\tNumpy  5.00
477  Qt\tООП\tC++\tMS Visual Studio\tРазработка ком...  6.00
478  SCALA\tSQL\tМатематический анализ\tСтатистичес...  7.00
479  Python\tLinux\tKubernetes\tAWS\tGit\tPeople Ma...  6.00
480                         среднее количество навыков  6.41

[481 rows x 2 columns]

Посмотрим дополнительно  какая средняя заработная плата указывается в вакансиях для  DS.

In [135]:
query = f''' with x as (select
v.key_skills, ((length(v.key_skills)-length(replace(v.key_skills,CHR(9),'')))::integer +1) sk_3, round(avg(coalesce(((v.salary_from+v.salary_to)/2),v.salary_from,v.salary_to)),0) avg_salary_total
from vacancies v
where(v.name ilike '%data scientist%' 
or v.name ilike '%data science%' 
or v.name ilike '%data science%'
 or v.name ilike '%исследователь данных%'
 or( v.name like '%ML%' and v.name not like '%HTML%')
 or v.name ilike '%machine learning%'
 or v.name ilike '%машинн%обучен%')
 group by v.key_skills)
 select
 *
 from x
 union  all
select
 'среднее значение',round(avg(x.sk_3),2),round(avg( x.avg_salary_total),2)
from x
where x.avg_salary_total <>0 and x.sk_3 <>0
    
'''
df = pd.read_sql_query(query, connection)
df  

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\2420455812.py:22: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


key_skills   sk_3   
0    Agile Project Management\tУправление проектами...   5.00  \
1    Big Data\tPython\tMachine Learning\tNLP\tData ...   6.00   
2    Big Data\tR&D\tResearch methodologies\tPh.D\tP...  14.00   
3    BPMN\tARIS\tАналитическое мышление\tБизнес-ана...  12.00   
4    C/C++\tLinux\tPython\tMachine learning\tнейрон...   5.00   
..                                                 ...    ...   
315  Статистический анализ\tАналитическое мышление\...   5.00   
316                               Управление проектами   1.00   
317  Управление проектами\tБазы данных\tData Analys...   5.00   
318                                               None    NaN   
319                                   среднее значение   6.51   

     avg_salary_total  
0                 NaN  
1                 NaN  
2                 NaN  
3                 NaN  
4                 NaN  
..                ...  
315               NaN  
316               NaN  
317               NaN  
318          97500.00  
319         180375.49  

[320 rows x 3 columns]

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [136]:
query = f''' select
v.experience опыт, round(avg(coalesce(((v.salary_from+v.salary_to)/2),v.salary_from,v.salary_to)),0) средняя_ЗП
from vacancies v
where(v.name ilike '%data scientist%' 
or v.name ilike '%data science%' 
or v.name ilike '%data science%'
 or v.name ilike '%исследователь данных%'
 or( v.name like '%ML%' and v.name not like '%HTML%')
 or v.name ilike '%machine learning%'
or v.name ilike '%машинн%обучен%')
 and (v.salary_from is not null or v.salary_to is not null)
 group by v.experience
'''
df = pd.read_sql_query(query, connection)
df  # текст запроса

C:\Users\Семья\AppData\Local\Temp\ipykernel_8136\606356926.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


опыт  средняя_ЗП
0           Нет опыта     74643.0
1  От 1 года до 3 лет    139675.0
2       От 3 до 6 лет    243115.0

***

Из 49 197 вакансий 1771 вакансия имеет отношение к данным, что составляет 3.6% от всех вакансий в базе.  1231  (2.5%) вакансии к продажам,  97 к бухгалтерии,74 для экономистов, 9  вакансий для юристов. Ранее очень популярные вакансии такие как экономист, бухгалтер и юрист практически отсутствуют. (Есть у меня сомнения, по количеству вакансий для юристов, скорее не верный отбор. Поправьте меня в резюме проекта, пожалуйста.) Для начинающего DS вакансий в базе всего 51 , что составляет 0.1 % от всех имеющихся в базе вакансий. Найти работу начинающему DS очень тяжело. Из 480 вакансий для DS  201 в которых в качестве ключевого навыка указан SQL или postgres, что составляет 42% , следовательно такие специалисты востребованы. Специалисов со знанием Питона запрашивают в 351 вакансиях, что составляет 73% от всех вакансий для DS, следовательно знание и умение писать коды на Питоне очень важный навык. 
Среднее количество навыков для DS составляет 6,51 , средняя заработная плата  в вакансиях с указанными навыками 180375,49 рублей. При анализе средней заработной платы в зависимости от опыта видно, что с большим опытом предлагается большая заработная плата, но если смотреть по количеству вакансий , то большинство требуются с опытом от 1 до 3 лет.

# Общий вывод по проекту

В настоящей базе представлено 49197 вакансий , 23501 работодатель из 1362 регионов с 294 сферами деятельности компаний, 593 региона не разместили не одной вакансии. Лидерами по количеству размещенных вакансий являются Москва, Санкт-Петербург, Минск, Новосибирск, Алматы. У  24073 вакансий указан уровень заработной платы, что  составляет 49% от всех представленных вакансий. Если  рассмотреть вакансии по критерию типа рабочего графика  то для соискателей на полный день из 36424 вакансй 52% без указанной заработной платы,  для соискателей  на удаленную работу из 9311 вакансий 49% без указанной заработной платы, для соискателей с гибким графиком из 2039 вакансий 58% без указанной заработной платы, для соискателей на вахту из 369 вакансий 27% без указанной заработной платы, для соискателей со сменным графиком из 1054 вакансий 22% без указанной заработной платы. Одной из причин не размещения уровня заработной платы работодателем может являться  открытие новой вакансии по новому направлению работы, связано это с тем, что работодатель еще  сам не определился с суммой оплаты. Также работодатель может не указывать желаемый  уровень заработной платы для сохранения возможности переговоров по уровню оплаты исходя из опыта, квалификации и результатов собеседования. Еще одной из причин может являться не желание показывать уровни оплаты сотрудников из за конкурентной среды среди работодателей. Также может привлечь на вакансию более опытного специалиста, чем ожидалось. В нашем случае для вакансий со сменным графиком менее всего пустых значений по уровню оплаты. Сменные графики чаще встречаются на производстве либо на предприятиях оказывающих услуги.   Проведя анализ вакансий со сменным графиком и с указанной заработной платой  выяснилось, что максимальный уровень оплаты  предлагается разработчикам Питона, консультантам по сопровождению 1С и продавцу - консультанту мебели Panda Terra(посмотрев сайт компании, на сегодняшний день есть вакансии  производственных рабочих со средним заработком 55-60 тысяч рублей).  Для более глубокого анализа в этом направлении можно посмотреть какие работодатели из каких регионов предлагают вакансии с максимальной заработной платой. И сделать выводы про регион.
Среднее значение  максимальной предлагаемой заработной платы  по всем вакансиям в базе 110537 тысяч рублей  среднее минимальное  71065 тысяч рублей. 
При условии анализа вакансий где указаны оба поля уровня оплаты лидерами по предлагаемой максимальной заработной плате являются  Алматы, Москва, Сочи, Санкт-Петербург и аутсейдерами являются  Челябинск, Воронеж, Смоленск, Уфа, но для окончательных выводов нужен более детальный анализ. Средняя максимальная заработная плата в этих городах (не беря в расчет вакансии где поля минимальной и максимальной заработной платы не заполнены) составила:
в Москве 169 929 рубль
в Сочи 152 180 рублей
в Санкт-Петербурге 141 734 рублей
в Алматы 111 835 рублей
В результате исследования выяснилось, что средняя максимальная заработная плата в Алматы ниже Москвы, Сочи и Санкт-Петербурга, что подтверждает, что выводы по максимуму  и минимуму делать не корректно, нужно более детально исследовать, смотреть среднюю, с учетом нулевых значений.
Найдем среднюю заработную плату в вышеназванных городах( с критериями исследования указанными выше):
в Москве 138 728.0 рубль
в Сочи 122 843 рублей
в Санкт-Петербурге 115 242.0 рублей
в Алматы 88 738.0 рублей
рейтинг средних заработных плат совпал со средними максимальными.
Для справки сделала запрос для получения информации по средней  оплате в городе Перми,  в результате  средняя заработная плата  составила 106 841.0, могу сказать по своему предприятию - у нас ниже даже нижней границы. Самая высокооплачиваемая в нашем городе вакансия по данным базы HH Product manager LATOKEN.
Самым востребованными условиями трудоустройства  являются полная занятость и полный день, количество вакансий с такими требованиями составило 35367 ( 72%), далее идут полная занятость и удаленная работа 7802 вакансии (16%) ,полная  занятость с гибким графиком 1593 (3%),  частичная занятость с удаленной работой 1312(3%) и пятерку лидеров замыкает полная занятость со сменным графиком 940 вакансий что составило 2% от всех размещенных на HH. Дополнительно исследовала пары с удаленной работой, так как по моему мнению ( очень субъективное) это самой удобный тип рабочего графика. В результате  в базе найдено 4 пары с удаленной работой : 
Полная занятость	Удаленная работа	7802 вакансий
               Частичная занятость	Удаленная работа	1312 вакансии
               Проектная работа	Удаленная работа	133 вакансии
                Стажировка	Удаленная работа	64 вакансии
Анализ количество вакансий в разрезе графиков работы:
              всего с графиком:	Полный день	36424.0
	всего с графиком:	Удаленная работа	9311.0
	всего с графиком:	Гибкий график	2039.0
	всего с графиком:	Сменный график	1054.0
	всего с графиком:	Вахтовый метод	369.
График с удаленной работой стоит на втором месте по количеству вакансий  что составляет 19% от всех вакансий, что говорит о популярности такого типа трудоустройства. 
Наиболее востребованным опытом  работы является период от года до трех лет, уровень заработной платы у такой категории вакансий ниже, чем у вакансий  от 3 до 6 лет и более 6 лет. Работодатели предпочитают взять на работу сотрудника с минимальным опытом, но в то же время имеющим навыки работы, чем обучать не опытного сотрудника или платить боле высокую заработную плату более  опытному.
Топ-5 работодателей из существующей базы : Яндекс, Ростелеком, Тинькофф, СБЕР и  Газпром нефть. При рассмотрении сфер деятельности этих компаний, только Газпром нефть является энергетическим предприятием. Остальные компании из списка относятся к  интернет компаниям, финансовым компаниям, компаниям из сферы IT.  Хоть и все они представляют разные отрасли, но являются ключевыми игроками на рынке и оказывают влияние на экономику России.
1148 работодателей в базе не указали вакансии это может говорить и о том, что эти работодатели уже нашли сотрудников как и то, что при заполнении анкеты не верно указали данные или приостановили деятельность.
Лидером среди компаний по масштабу географии поиска сотрудников является компания Яндекс.  Они разместили вакансии в 181 городе. На втором месте Ростелеком – 152 региона и на третьем Спецремонт -116 регионов.  Проанализировав топ 20   сфер деятельности этих компаний выяснилось, что основном  это компании связи , но есть и те, что не указали сферу  своей деятельности. Всего в базе 8419 работодателей, кто не указал сферу деятельности. Некоторые компании не указывают сферу деятельности из соображений конфиденциальности, по причине разработки нового продукта , который преждевременно афишировать. По причине  множества разнообразных сфер деятельности. По причине  привлечения сотрудников из разных сфер, чтобы не ограничивать соискателей указанными сферами. Причины могут быть разные, вплоть до не желания заполнить это графу кадровым работником, подающим объявление.  Топ- 5 регионов, где не указали сферу деятельности : Москва, Санкт-Петербург, Минск, Екатеринбург и Новосибирск. 
При исследовании количества сфер деятельности, которое указывают компании, выяснилось, что самое большое количество сфер деятельности указанное работодателями пять. Компания с самым большим количеством вакансий - Яндекс с одной сферой деятельности, в то же время Ростелеком и Тинкоф указали по 4 и 5 сфер деятельности.
Из 49197 вакансий самый большую долю   сфер деятельности  занимает "Разработка программного обеспечения", что  составляет 7%,  далее идут системная интеграция, автоматизации технологических и бизнес-процессов предприятия, ИТ., затем итнергет-компания и так далее. Преимущественное большинство компаний публикующее вакансии  из сферы ИТ.
 Из 49 197 вакансий 1771 вакансия имеет отношение к данным, что составляет 3.6% от всех вакансий в базе.  1231  (2.%) вакансии к продажам,  97 к бухгалтерии,74 для экономистов, 9  вакансий для юристов. Ранее очень популярные вакансии такие как экономист, бухгалтер и юрист практически отсутствуют.
Количество вакансий  для начинающего DS  в базе всего 51, что составляет 0.1 % от всех имеющихся в базе вакансий. Найти работу начинающему DS очень тяжело. Из 480 вакансий для DS 201 в которых в качестве ключевого навыка указан SQL или postgres, что составляет 42% , следовательно такие специалисты востребованы. От DS требуется в среднем 6-7 навыков,  средняя заработная плата  в вакансиях с указанными навыками 180375,49 рублей, что выше  среднего значения максимальной предлагаемой заработной платы  по всем вакансиям в базе (110537 тысяч рублей).  Запрос  зависимости средней заработной платы DS  от опыта работы  показал, что :
Без  опыта-74 643.0 рублей,
От 1 года до 3 лет-139 675.0 рублей
От 3 до 6 лет-243 115.0 рублей
Закономерно, что с большим опытом предлагается большая заработная плата, но если смотреть по количеству вакансий, то более популярен опыт от 1 до 3 лет.


In [65]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования